Offline Partial Eval

ncough opened the pull request:
Wrapper around existing online partial evaluation work to perform offline partial evaluation. Rather than producing the semantics for a single instruction opcode, this approach produces a lifter for some subset of the architecture.

The process looks roughly like the following:
  1. Convert the decoder & instruction encoding structures into ASL functions
  2. Identify all of the reachable functions from this initial set up to a configurable frontier
  3. Simplify these functions to remove annoying/unsupported features
  4. Transform each function such that loop bounds and bitvector widths are always known
  5. Run the existing online partial evaluation over each instruction encoding function
  6. Do some cleanup and pruning based on the results of the partial evaluation
  7. Run a taint analysis over the evaluated instruction encoding functions to identify
     what is known at lifttime vs runtime
  8. Transform all runtime influenced statements and expressions into IR generating operations
     against an assumed interface
  9. Print the result as an OCaml program with the IR interface instantiated as ASL IR

The above works for a significant portion of AArch64 matching the existing lifters coverage for all but `aarch64_memory_vector_single_post_inc`, `aarch64_memory_vector_single_no_wb`, `aarch64_memory_vector_multiple_post_inc`, `aarch64_memory_vector_multiple_no_wb`. However, the resulting OCaml program is ~200k lines long and takes ~5min to build. This can mostly be attributed to some aggressive specialisation in stage 4 and the resulting unrolling transforms introduced by stage 5.

There are still a decent number of things to do:
  - Improvements to AArch64 coverage:
    - LExpr BitTuple support
  - Improvements to compile time / size:
    - Break the produced OCaml program into a series of files, rather than one
    - Support symbolic for loops throughout the pipeline, to reduce unrolling and specialisation
    - Simplify based on branch conditions (e.g. `if X = '1' then c` should replace X with '1' in c)
  - 'Quality' of resulting program:
    - Comparison of online & offline lifter results
    - Implement a form of copy prop over the IR generating operations produced in stage 8
    - Implement some trivial identities within the IR interface
 
Given the compile time issues, this PR includes a generate lifter that supports no instructions.
A new lifter can be generated using `echo ':gen A64 aarch64.+' | dune exec asli` from the project directory.
A subsequent `dune build` will compile the new lifter.
There are two entry points to the lifter:
  - `asloff-sem OPCODE` will dump the semantics of OPCODE
  - `asloff-coverage PAT` will run the coverage test for all instruction encodings matching PAT

ncough committed:
Fix a typo and stub out some complex behaviours in spec

ncough committed:
Add bool prims to prim list

ncough committed:
Add missing prims for type inference in transforms

ncough committed:
StatefulIntToBit fixes

Correct visit order bug in StatefulIntToBit, support
register type

ncough committed:
Pass to remove temp dynamic bitvectors

Some common idioms can be easily rewritten, avoiding
complexity udring dis. Currently unused.

ncough committed:
Rework mask support to reduce to bitops

Masks are never symbolic, so they can be trivially
rewritten as a masking and equiv test if needed.

ncough committed:
Centralise no_inline sets

May want to make this a configuration option to dis

ncough committed:
Extend stmt append to halt on terminal stmts

Cuts of prefix early, even if prefix does not end with terminal

ncough committed:
Symbolic array accesses

Assumes no mixed use of array indices

ncough committed:
Some helpers for dealing with fn signatures

ncough committed:
Mostly working offline partial eval pipeline

Capable of extracting and simplifying behaviours for
most of the instruction encodings on AArch64.
Offline gen is missing specs for primitives as well
as the prelude required for the ocaml backend.

ncough committed:
Build decoder tests without patterns or early returns

ncough committed:
Functionality to get the type of an expression post-dis

ncough committed:
Add entry point, coverage testing for offline

ncough committed:
Add a pass to replace register types with bits

ncough committed:
Add req transform for cases

ncough committed:
Extend offline to integer operations

ncough committed:
Add coverage script for offline

Some differences in exception printing, unclear

ncough committed:
Hack in prj file support to offline coverage, matching existing results

ncough committed:
Reduce debug printing

ncough committed:
Fixes for branches

ncough committed:
Fixes for memory operations

ncough committed:
Prevent large vector ops for now

ncough committed:
Partially extend req patterns

ncough committed:
Extend symbolic to support reducing HiLo widths

ncough committed:
Coverage fixes

ncough committed:
Consider producing many individual files

ncough committed:
Add getlast util

ncough committed:
Extend offline to support floats

ncough committed:
Clean out redundant transforms, filters

ncough committed:
Fix IntToBit for cvt_int_bits

If feeding in a larger bit rep, its necessary to slice down

ncough committed:
Prune useless globals

ncough committed:
Generalise req patterns to capture generic case & LowestSetBit

ncough committed:
Improve sub_int simplification

Given (e - (a + b)) apply existing rules as ((e - a) - b)

ncough committed:
Rather than failing on unsupported PSTATE mods, insert assert

ncough committed:
Ocaml backend cleanup

ncough committed:
Prune unsupported impdefs, block insts with unsupported state access

ncough committed:
Offline lifter cleanup, push to all of aarch64

ncough committed:
Collect TODOs

ncough committed:
Cleanup build, bundle empty implementation to avoid costly builds

ncough committed:
Fixups for coverage

Don't remove dead globals to maintain coverage results
Set CRC support

ncough committed:
Convert tuple return from FPToFixedJS to bitvec

ncough committed:
Offline transform cleanup

ncough committed:
Replace bool impdefs with false rather than assert

ncough committed:
Add missing fp op

ncough committed:
Fix sed usage

ncough committed:
Fix coverage for FPToFixedJS

We no longer have any stage 2,3 or 4 failures!

katrinafyi reviewed (approved):
Amazing!! You can merge this whenever you're ready, https://github.com/UQ-PAC/aslp/pull/53 can merge cleanly on top of this.

ncough committed:
Extend ocaml backend to produce many files

ncough committed:
Eliminate unnecessary temps in offline transform

The naive offline transform treats all runtime variables
as temporaries that must be declared in the runtime program.
Often these temporaries simply hold intermediate calculations
and can be trivially removed through a copy prop pass.

As we do not want to introduce a copy prop into the generated
lifter, we can mimic its effects over the produced lifter by
identifying candidate temporaries and changing them to directly
hold the intermediate calculations, rather than representing
runtime variables.

ncough commented:
Apologies, made a few improvements. After splitting the OCaml program across multiple files, build times are now down to ~1min. Also managed to get the copy prop pass implemented. Will merge once the tests pass.
